kad-dht/README.md

```diff
@@ -68,6 +68,31 @@ The concurrency of node and value lookups are limited by parameter `α`, with a
 default value of 3. This implies that each lookup process can perform no more
 than 3 inflight requests, at any given time.
 
+## Client and server mode
+
+When the libp2p Kademlia protocol is run on top of a network of heterogeneous
+nodes, unrestricted nodes should operate in _server mode_ and restricted nodes,
+e.g. those with intermittent availability, high latency, low bandwidth, low
+CPU/RAM/Storage, etc., should operate in _client mode_.
+
+As an example, running the libp2p Kademlia protocol on top of the Internet,
+publicly routable nodes, e.g. servers in a datacenter, would operate in _server
+mode_ and non-publicly routable nodes, e.g. laptops behind a NAT and firewall,
+would operate in _client mode_.
```

Comment:
This seems pretty good, although the advice here is all sort of fuzzy and situational. I'm adding some context in case you think it's reasonable to rephrase.

The definition of intermittent availability depends on the parameters of a given kad-dht network (replication factor, republishing periods, record expiration times, etc.). For example, having a laptop with UPnP support behind a NAT that's open for a few hours consecutively is probably a fine DHT server for networks with replication factors high enough to accommodate that (e.g. tuned to assume many nodes disappear after an hour).

The key is basically that nodes should be in client mode if they think they'll ultimately be a drag on the network (e.g. everyone behind a symmetric NAT or firewall), but what that means is up to them. Most implementations should probably do their best to figure this out probabilistically/automatically so that normal users don't have to consider this.

Reply:
I'm not sure if this is covered elsewhere in the spec, but networks may also want to add their own rules for routing table membership to exclude bad peers (e.g. those without public IP addresses) so they're not just relying on users correctly setting client mode. However, if using rules they should probably be as symmetric/uniformly applicable as possible to maintain the coherence of the network's routing tables.

Reply:
https://github.com/libp2p/specs/pull/392/commits/710c4998afa02a07aa871492d01b5f5f9f2f70c6 adds the factors above as examples to take into consideration. I am not sure how to make it less fuzzy without loosing generality and thus not being applicable to all use-cases. Suggestions welcome.

Reply:
> I'm not sure if this is covered elsewhere in the spec, but networks may also want to add their own rules for routing table membership to exclude bad peers (e.g. those without public IP addresses) so they're not just relying on users correctly setting client mode. However, if using rules they should probably be as symmetric/uniformly applicable as possible to maintain the coherence of the network's routing tables.

Oh, just saw this. I don't think this is covered anywhere. Agreed. I am not sure it needs to be documented though.